CMakeLists.txt

```diff
@@ -948,7 +948,24 @@ if(CONFIG_USERSPACE)
 endif()
 
 get_property(CSTD GLOBAL PROPERTY CSTD)
-set_ifndef(CSTD c99)
+if(NOT DEFINED CSTD)
+  if(CONFIG_STD_C23)
+    set(CSTD c2x)
+  elseif(CONFIG_STD_C17)
+    set(CSTD c17)
+  elseif(CONFIG_STD_C11)
+    set(CSTD c11)
+  elseif(CONFIG_STD_C99)
+    set(CSTD c99)
+  elseif(CONFIG_STD_C90)
+    set(CSTD c90)
+  else()
+    assert(0 "Unreachable code. Expected C standard to have been chosen.")
+  endif()
+else()
+  message(DEPRECATION
+    "Global CSTD property is deprecated, see Kconfig.zephyr for C Standard options.")
```

Comment:
Technically, this is deprecating not just the global property, but also passing `-DCSTD=...` as seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/fd1a129a3e6330bdf44324e47ac7638dd6cdf5a5/tests/subsys/bindesc/definition/testcase.yaml#L24-L35

Reply:
> should we introduce a `CONFIG_STD_C_EXTENSIONS` too for the gnu extensions?

Sure, if @tejlmand agrees.

I'm not too sure how C standard selection should work, because I'm only learning now that CMake lets you control this on a per-target basis, using [C_STANDARD](https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD.html) and [C_EXTENSIONS](https://cmake.org/cmake/help/latest/prop_tgt/C_EXTENSIONS.html) properties. Maybe a global setting is not needed?

Reply:
I have to admit, this is kind of getting over my head 🙃
I tried with `C_STANDARD` and `C_EXTENSIONS` but this doesn't make the `-std=` compiler argument explicit it seems, and fails in some tests.